python/tk_multi_publish2/api/graph.py

```diff
@@ -0,0 +1,588 @@
+
+        if item == self.root_item:
+            raise sgtk.TankError("Removing the root item is not allowed.")
+
+        self._remove_items([item])
+
+    def save(self, path):
+        """Save the graph to disk at the supplied path."""
+
+        with open(path, "wb") as graph_file:
+            try:
+                pickle.dump(self, graph_file, protocol=pickle.HIGHEST_PROTOCOL)
+            except Exception:
+                import traceback
+                logger.debug(traceback.format_exc())
+                raise
+
+    @property
+    def items(self):
+        """Returns a list of all items in the graph. Order is not guaranteed."""
+        return self._items
+
+    @property
+    def persistent_items(self):
+        """Returns a list of persistent items in the graph."""
+        return [i for i in self.root_item.children if i.persistent]
+
+    @property
+    def root_item(self):
+        """Returns the root item of this graph."""
+        return self._root_item
+
+    def _init_state(self):
+        """The init/reset the graph state."""
+
+        # NOTE: this is factored out to simplify a full reset of the object.
+
+        # these internal members define the structure and relationships of the
+        # graph itself. we intentionally only store complex objects in a single
+        # place like _items, _tasks, and _plugins. All other lookups are id
+        # based to keep the complexity down and simplify serialization of the
+        # graph.
+        self._child_lookup = {}
+        self._items = []
+        self._items_by_id = {}
+        self._parent_lookup = {}
+        self._plugins = {}
+        self._tasks = []
+        self._tasks_by_item = {}
+        self._tasks_by_id = {}
+        self._items_by_task = {}
+
+        # The root item is the sole parent of all top-level publish items. It
+        # has no use other than organization and provides an easy interface for
+        # beginning iteration and accessing all top-level items.
+        self._root_item = PublishItem(self, "root", "_root", "_root")
+        self.add_item(self._root_item)
+
+    def _format_graph(self, parent_item, depth=0):
+        """
+        Depth first traversal and string formatting of the graph given a root
+        node.
+
+        :param parent_item: The item to begin with
+        :param depth: The current depth of the traversal (used for indentation)
+        """
+
+        graph_str = ""
+
+        for item in self.get_children(parent_item):
+
+            # format the item itself
+            graph_str += "%s[item] %s\n" % ((depth * 2) * " ", str(item))
+
+            # now do the item's tasks
+            for task in item.tasks:
+                graph_str += "%s[task] %s\n" % (((depth * 2) + 2) * " ", str(task))
+
+            # process the children as well
+            graph_str += "%s" % (self._format_graph(item, depth=depth+1),)
+
+        return graph_str
+
+    def _remove_items(self, items):
+        """
```

Comment:
This is the method that makes me wonder if this is the right approach. The internal state of the object is super simple and the lookups are really easy to manage when it comes to traversing, adding items, and querying relationships, but removing items is a bit complex.

Reply:
Storing the dependencies as a three when persisting will be fine I think.

As for ids, you don't need to build them all up front. You can probably use the Python object's id as the model id. Or you can create your own mapping of id's and weakrefs to the item/task.

Reply:
I think we let worries about serialization be an argument for a less DX-friendly API. Serialization and the design of it can help motivate a certain approach, but i definitely think we should start by looking at "what do we want the TD experience to be like" and then figure out what the internals look like.

I had similar thoughts to JFs', but manage to lose my comments as a accidentally cancelled my CR (github was a little bit confusing if you are trying to dismiss the submit UI for CR!)